module/zfs/dnode.c

```diff
@@ -822,19 +821,14 @@ dnode_move_impl(dnode_t *odn, dnode_t *ndn)
 	ndn->dn_newgid = odn->dn_newgid;
 	ndn->dn_newprojid = odn->dn_newprojid;
 	ndn->dn_id_flags = odn->dn_id_flags;
-	dmu_zfetch_init(&ndn->dn_zfetch, NULL);
-	list_move_tail(&ndn->dn_zfetch.zf_stream, &odn->dn_zfetch.zf_stream);
-	ndn->dn_zfetch.zf_dnode = odn->dn_zfetch.zf_dnode;
```

Comment:
Does this leak the zstream_t's on the zf_stream?

The move callbacks are pretty tricky code to get right.  IMO, if we aren't able to maintain the move callbacks, because we can't test them, because they are dead code, we should remove them from the codebase.

Reply:
dmu_zfetch_fini(&odn->dn_zfetch) below should free the old streams gracefully.

Reply:
OK.  I'm fine with integrating this change, but I would welcome a future PR to remove dnode_move_impl() entirely.